src/login/logind-session-device.c

```diff
@@ -417,14 +417,20 @@ void session_device_free(SessionDevice *sd) {
         assert(sd);
 
         if (sd->pushed_fd) {
+                _cleanup_free_ char *fdname = NULL;
                 const char *m;
+                int r;
 
                 /* Make sure to remove the pushed fd. */
 
-                m = strjoina("FDSTOREREMOVE=1\n"
-                             "FDNAME=session-", sd->session->id);
+                r = asprintf(&fdname, "device-%u-%u-session-%s",
+                                      major(sd->dev), minor(sd->dev), sd->session->id);
+                if (r < 0) {
```

Comment:
Hmm, `r < 0`? I think you mean `r >= 0`.

I don't think it makes much sense to do asprintf and then strjoina. Just asprintf the whole thing.

Reply:
Or do
```C
if () {
    char buf[strlen("FDSTOREREMOVE=1\n" "FDNAME=session--device--") + strlen(sd->session->id) + 2*DECIMAL_STR_MAX(unsigned) + 1];
    ...
    xsprintf(buf, ...);
```
This will avoid the ugly error handling of failing asprintf.